module/scheduler.cpp

```diff
@@ -555,8 +617,94 @@ void Scheduler::SendToLockManager(const TransactionHolder* txn_holder) {
   }
 }
 
-void Scheduler::AbortTransaction(const TransactionHolder* txn_holder) {
-  // TODO
+void Scheduler::AbortTransaction(TransactionHolder* txn_holder, bool was_dispatched) {
+  CHECK(txn_holder->InvolvedPartitions().size() == 1) << "MP not impelemented";
+
+  auto txn = txn_holder->GetTransaction();
+  auto txn_id = txn->internal().id();
+  LOG(INFO) << "Aborting txn " << txn_id;
+  switch (txn->internal().type()) {
+    case TransactionType::SINGLE_HOME: {
+      txn->set_status(TransactionStatus::ABORTED);
+      SendToCoordinatingServer(txn_holder);
+      break;
+    }
+    case TransactionType::MULTI_HOME: {
+      txn->set_status(TransactionStatus::ABORTED);
+      if (was_dispatched) {
+        // LOs have already been deleted
+        SendToCoordinatingServer(txn_holder);
+      } else {
```

Comment:
You can check dispatch status of a txn with `GetWorker()` on a holder. If the worker is empty then it is not yet dispatched. Also, this code can be simplified to:
```
SendToCoordinatingServer(txn_holder);
if (<was_not_dispatched>) {
....  // Remove the other SendToCoordinatingServer below
}
```